packages/mdc-menu-surface/foundation.js

```diff
@@ -0,0 +1,572 @@
+  }
+
+  /**
+   * @param {!Corner} corner Origin corner of the menu surface.
+   * @return {number} Vertical offset of menu surface origin corner from corresponding anchor corner.
+   * @private
+   */
+  getVerticalOriginOffset_(corner) {
+    const {anchorHeight} = this.measures_;
+    const isBottomAligned = Boolean(corner & CornerBit.BOTTOM);
+    const avoidVerticalOverlap = Boolean(this.anchorCorner_ & CornerBit.BOTTOM);
+    let y = 0;
+
+    if (isBottomAligned) {
+      y = avoidVerticalOverlap ? anchorHeight - this.anchorMargin_.top : -this.anchorMargin_.bottom;
+    } else {
+      y = avoidVerticalOverlap ? (anchorHeight + this.anchorMargin_.bottom) : this.anchorMargin_.top;
+    }
+    return y;
+  }
+
+  /**
+   * @param {!Corner} corner Origin corner of the menu surface.
+   * @return {number} Maximum height of the menu surface, based on available space. 0 indicates should not be set.
+   * @private
+   */
+  getMenuSurfaceMaxHeight_(corner) {
+    let maxHeight = 0;
+    const {viewportDistance} = this.measures_;
+    const isBottomAligned = Boolean(corner & CornerBit.BOTTOM);
+    const {MARGIN_TO_EDGE} = MDCMenuSurfaceFoundation.numbers;
+
+    // When maximum height is not specified, it is handled from css.
+    if (isBottomAligned) {
+      maxHeight = viewportDistance.top + this.anchorMargin_.top - MARGIN_TO_EDGE;
+      if (!(this.anchorCorner_ & CornerBit.BOTTOM)) {
+        maxHeight += this.measures_.anchorHeight;
+      }
+    } else {
+      maxHeight = viewportDistance.bottom - this.anchorMargin_.bottom + this.measures_.anchorHeight - MARGIN_TO_EDGE;
+      if (this.anchorCorner_ & CornerBit.BOTTOM) {
+        maxHeight -= this.measures_.anchorHeight;
+      }
+    }
+
+    return maxHeight;
+  }
+
+  /** @private */
+  autoPosition_() {
+    // Compute measurements for autoposition methods reuse.
+    this.measures_ = this.getAutoLayoutMeasurements_();
+
+    const corner = this.getOriginCorner_();
+    const maxMenuSurfaceHeight = this.getMenuSurfaceMaxHeight_(corner);
+    const verticalAlignment = (corner & CornerBit.BOTTOM) ? 'bottom' : 'top';
+    let horizontalAlignment = (corner & CornerBit.RIGHT) ? 'right' : 'left';
+    const horizontalOffset = this.getHorizontalOriginOffset_(corner);
+    const verticalOffset = this.getVerticalOriginOffset_(corner);
+    let position = {
+      [horizontalAlignment]: horizontalOffset ? horizontalOffset : '0',
+      [verticalAlignment]: verticalOffset ? verticalOffset : '0',
+    };
+    const {anchorWidth, surfaceWidth} = this.measures_;
+    // Center align when anchor width is comparable or greater than menu surface, otherwise keep corner.
+    if (anchorWidth / surfaceWidth > numbers.ANCHOR_TO_MENU_SURFACE_WIDTH_RATIO) {
+      horizontalAlignment = 'center';
+    }
+
+    // If the menu-surface has been hoisted to the body, it's no longer relative to the anchor element
```

Comment:
I noticed that this method used to bail out if there was no anchor element but it no longer does that. What are the implications of that on this method?

Reply:
The menu-surface can work without an anchor element (e.g. context menus). If the menu is opened without an anchor element and without having a position, the menu will open at `x: 0`/`y: 0`.